src/inertial_sense_ros.cpp

```diff
@@ -5,857 +5,1227 @@
+            ixEuler eul = {M_PI, 0, 0.5 * M_PI};
+            // ENU-to-NED quaternion
+            euler2quat(eul, qn2enu);
+            // NED-to-body quaternion
+            mul_Quat_ConjQuat(qn2b, qe2b, qe2n);
+            // ENU-to-body quaternion
+            mul_Quat_ConjQuat(qenu2b, qn2b, qn2enu);
+            // ECEF-to-ENU quaternion
+            mul_Quat_Quat(qe2enu, qn2enu, qe2n);
+            // Body-to-ENU rotation matrix
+            rotMatB2R(qenu2b, Rb2enu);
+            // ECEF-to-ENU rotation matrix
+            rotMatB2R(qe2enu, buf);
+            transpose_Mat3(Re2enu, buf);
+
+            // Pose
+            // Transform position from ECEF to ENU and attitude from body to ENU
+            transform_6x6_covariance(Pout, poseCov, Re2enu, Rb2enu);
+            for (int i = 0; i < 36; i++) 
+            {
+                enu_odom_msg.pose.covariance[i] = Pout[i];
+            }
+            // Twist
+            // Transform velocity from ECEF to ENU and angular rate from body to ENU
+            transform_6x6_covariance(Pout, twistCov, Re2enu, Rb2enu);
+            for (int i = 0; i < 36; i++) 
+            {
+                enu_odom_msg.twist.covariance[i] = Pout[i];
+            }
+            
+            enu_odom_msg.header.stamp = ros_time_from_week_and_tow(msg->week, msg->timeOfWeek);
+            enu_odom_msg.header.frame_id = frame_id_;
+
+            // Position
+                //Calculate in NED then convert
+            ixVector3d llaPosRadians;
+                //ecef to lla (rad,rad,m)
+            ecef2lla(msg->ecef, llaPosRadians, 5);
+            ixVector3 ned;
+            ixVector3d refLlaRadians;
+                //convert refLla_ to radians
+            lla_Deg2Rad_d(refLlaRadians, refLla_);
+                //lla to ned
+            lla2ned_d(refLlaRadians, llaPosRadians, ned);
+
+            // Rearrange from NED to ENU
+            enu_odom_msg.pose.pose.position.x = ned[1];
+            enu_odom_msg.pose.pose.position.y = ned[0];
+            enu_odom_msg.pose.pose.position.z = -ned[2];
+
+            // Attitude
+
+            enu_odom_msg.pose.pose.orientation.w = qenu2b[0];
+            enu_odom_msg.pose.pose.orientation.x = qenu2b[1];
+            enu_odom_msg.pose.pose.orientation.y = qenu2b[2];
+            enu_odom_msg.pose.pose.orientation.z = qenu2b[3];
+
+            // Linear Velocity
+                //same as NED but rearranged.
+            ixVector3 result, theta;
+            quatConjRot(result, qe2n, msg->ve);            
+
+            enu_odom_msg.twist.twist.linear.x = result[1];
+            enu_odom_msg.twist.twist.linear.y = result[0];
+            enu_odom_msg.twist.twist.linear.z = -result[2];
+
+            // Angular Velocity
+
+            // Transform from body frame to ENU
+            ixVector3 angVelImu = {(f_t)imu_msg.angular_velocity.x, (f_t)imu_msg.angular_velocity.y, (f_t)imu_msg.angular_velocity.z};
+            quatConjRot(result, qenu2b, angVelImu);
```

Comment:
I think it should be quatRot() here (rotation from body to reference frame)

Reply:
OK